third-party/realdds/src/dds-stream-sensor-bridge.cpp

```diff
@@ -0,0 +1,436 @@
+            if( streaming_needed )
+                add_implicit_profiles();
+            commit();
+        }
+        catch( std::exception const & e )
+        {
+            std::string error_string
+                = "failure trying to start/stop '" + stream.server->name() + "': " + e.what();
+            LOG_ERROR( error_string );
+            if( _on_error )
+                _on_error( error_string );
+        }
+    }
+}
+
+
+bool profiles_are_compatible( std::shared_ptr< dds_stream_profile > const & p1,
+                              std::shared_ptr< dds_stream_profile > const & p2,
+                              bool any_format = false )
+{
+    auto vp1 = std::dynamic_pointer_cast< realdds::dds_video_stream_profile >( p1 );
+    auto vp2 = std::dynamic_pointer_cast< realdds::dds_video_stream_profile >( p2 );
+    if( ! ! vp1 != ! ! vp2 )
+        return false;  // types aren't the same
+    if( vp1 && vp2 )
+        if( vp1->width() != vp2->width() || vp1->height() != vp2->height() )
+            return false;
+    if( ! any_format && p1->format() != p2->format() )
+        return false;
+    return p1->frequency() == p2->frequency();
+}
+
+
+static std::shared_ptr< realdds::dds_stream_profile >
+find_profile( std::shared_ptr< realdds::dds_stream_server > const & stream,
+              std::shared_ptr< realdds::dds_stream_profile > const & profile,
+              bool any_format = false )
+{
+    auto & stream_profiles = stream->profiles();
+    auto it = std::find_if( stream_profiles.begin(),
+                            stream_profiles.end(),
+                            [&]( std::shared_ptr< realdds::dds_stream_profile > const & sp )
+                            { return profiles_are_compatible( sp, profile, any_format ); } );
+    std::shared_ptr< realdds::dds_stream_profile > found_profile;
+    if( it != stream_profiles.end() )
+        found_profile = *it;
+    return found_profile;
+}
+
+
+std::shared_ptr< realdds::dds_stream_server >
+dds_stream_sensor_bridge::find_server( std::shared_ptr< realdds::dds_stream_profile > const & profile )
+{
+    if( ! profile )
+        DDS_THROW( runtime_error, "null profile" );
+    
+    auto server = profile->stream();
+    if( ! server )
+        DDS_THROW( runtime_error, "profile " + profile->to_string() + " has no stream" );
+
+    auto sensorit = _sensors.find( server->sensor_name() );
+    if( sensorit == _sensors.end() )
+        DDS_THROW( runtime_error, "no sensor '" + server->sensor_name() + "'" );
+    auto & sensor = sensorit->second;
+
+    auto streamit = sensor.streams.find( server->name() );
+    if( streamit == sensor.streams.end() )
+        DDS_THROW( runtime_error, "no stream '" + server->name() + "'" );
+    auto & stream = streamit->second;
+
+    if( stream.server != server )
+        DDS_THROW( runtime_error, "profile stream '" + server->name() + "' does not match server" );
+    return stream.server;
```

Comment:
Isn't this just returning `profile->stream()`? Why do we need this function?

Reply:
Because it does a lot of checking.

Reply:
Then call it `check_profile_stream`

Reply:
I want whoever calls to use the return value and not the original server. So I'd rather keep this name. :)